src/libsystemd/sd-device/sd-device.c

```diff
@@ -415,6 +415,31 @@ _public_ int sd_device_new_from_stat_rdev(sd_device **ret, const struct stat *st
         return sd_device_new_from_devnum(ret, type, st->st_rdev);
 }
 
+_public_ int sd_device_new_from_devname(sd_device **ret, const char *devname) {
+        struct stat st;
+
+        assert_return(ret, -EINVAL);
+        assert_return(devname, -EINVAL);
+
+        if (isempty(path_startswith(devname, "/dev")))
+                return -EINVAL;
```

Comment:
So, I think we should tweak this one tiny bit more. i.e. we should use device_path_parse_major_minor() here, i.e. try to parse the major/minor directly from the path if the path is a /dev/block/maj:min path. 

Why bother? because it means sd_device_new_from_devname() actually can work for these paths even if udev isn't running, and hence no /dev/block/ + /dev/char/ symlink created yet.